source/common/conn_pool/conn_pool_base.h

```diff
@@ -73,24 +73,36 @@ class ActiveClient : public LinkedObject<ActiveClient>,
   // Returns the number of active streams on this connection.
   virtual uint32_t numActiveStreams() const PURE;
 
+  // Return true if it is ready to dispatch the next stream.
+  virtual bool readyForStream() const {
+    ASSERT(!supportsEarlyData());
+    return state_ == State::Ready;
+  }
+
   // This function is called onStreamClosed to see if there was a negative delta
   // and (if necessary) update associated bookkeeping.
   // HTTP/1 and TCP pools can not have negative delta so the default implementation simply returns
   // false. The HTTP/2 connection pool can have this state, so overrides this function.
   virtual bool hadNegativeDeltaOnStreamClosed() { return false; }
 
   enum class State {
-    Connecting, // Connection is not yet established.
-    Ready,      // Additional streams may be immediately dispatched to this connection.
-    Busy,       // Connection is at its concurrent stream limit.
-    Draining,   // No more streams can be dispatched to this connection, and it will be closed
-    // when all streams complete.
-    Closed // Connection is closed and object is queued for destruction.
+    Connecting,        // Connection is not yet established.
+    ReadyForEarlyData, // Any additional early data stream can be immediately dispatched to this
+                       // connection.
+    Ready,             // Additional streams may be immediately dispatched to this connection.
+    Busy,              // Connection is at its concurrent stream limit.
+    Draining,          // No more streams can be dispatched to this connection, and it will be
+                       // closed when all streams complete.
+    Closed             // Connection is closed and object is queued for destruction.
   };
 
   State state() const { return state_; }
 
   void setState(State state) {
+    if (state == State::ReadyForEarlyData && !supportsEarlyData()) {
+      IS_ENVOY_BUG("Non-early-data compliant client gets early data ready");
```

Comment:
nit: How about: Unable to set state to ReadyForEarlyData in a client which does not support early data.